unpackaged/vesuvio_calibration/calibration_scripts/calibrate_vesuvio_helper_functions.py

```diff
@@ -216,3 +187,116 @@ def generate_fit_function_header(function_type, error=False):
+
+    def filter_peak_centres_for_invalid_detectors(self, detector_range, peak_table):
+        """
+          Finds invalid detectors and filters the peak centres. Caches the invalid detectors found to avoid repeat identification.
+          @param detector_range detectors to consider, must be either FRONT or BACKSCATTERING range.
+          @param peak_table - name of table containing fitted parameters each spectra.
+
+          @returns peak_centres - a list of peak fitted peak centres, with those that represent invalid detectors marked nan.
+        """
+
+        invalid_detectors = self.identify_and_set_invalid_detectors_from_range(detector_range, peak_table)
+        peak_centres = EVSMiscFunctions.read_fitting_result_table_column(peak_table, 'f1.LorentzPos', detector_range)
+        peak_centres[invalid_detectors] = np.nan
+        return peak_centres
+
+    def identify_and_set_invalid_detectors_from_range(self, detector_range, peak_table):
+        """
+          Finds invalid detectors and caches the invalid detectors. Will not look to calculate if invalid detectors already exist.
+          @param detector_range detectors to consider, must be either FRONT or BACKSCATTERING range.
+          @param peak_table - name of table containing fitted parameters each spectra.
+
+          @returns invalid_detectors - a list of the index's of invalid detector, in the context of the range they belong to.
+        """
+
+        peak_centres = EVSMiscFunctions.read_fitting_result_table_column(peak_table, 'f1.LorentzPos', detector_range)
+        peak_centres_errors = EVSMiscFunctions.read_fitting_result_table_column(peak_table, 'f1.LorentzPos_Err', detector_range)
+
+        if detector_range == EVSGlobals.FRONTSCATTERING_RANGE:
+            if not self._detectors_preset and not self._invalid_detectors_front.any():
+                self._invalid_detectors_front = self._identify_invalid_spectra(peak_table, peak_centres, peak_centres_errors,
+                                                                               detector_range)
+                self._print_invalid_detectors(self._invalid_detectors_front, detector_range)
+            return self._invalid_detectors_front
+        elif detector_range == EVSGlobals.BACKSCATTERING_RANGE:
+            if not self._detectors_preset and not self._invalid_detectors_back.any():
+                self._invalid_detectors_back = self._identify_invalid_spectra(peak_table, peak_centres, peak_centres_errors,
+                                                                              detector_range)
+                self._print_invalid_detectors(self._invalid_detectors_back, detector_range)
+            return self._invalid_detectors_back
+        else:
+            raise AttributeError("Spec list invalid - must represent either front or back detectors.")
+
+    @staticmethod
+    def _print_invalid_detectors(invalid_detectors, detector_range):
```

Comment:
I've made this change - out of interest why do you prefer to avoid static members, I see conflicting advise regarding this?

Reply:
It is just a personal preference. The use cases of the static function are limited as it cannot access class members. The only common use case I think is factory functions. I think in most cases it is better to extract the function as an external function. In your scenario, since all arguments of the function are private variables and it is only used internally, it is better from an encapsulation perspective to hide the function. I'm also interested to hear your perspective.

Reply:
I don't have a strong opinion really. If I have a class member which does not access private member variables, if it is a function only likely to be used within that class I generally keep it within that class and mark it static. The static decorator just signals to the reader that it will not mutate any class members.